File: src/test/java/com/heybys/oddments/fooddelivery/service/shop/ShopServiceTest.java
package com.heybys.oddments.fooddelivery.service.shop;

import static java.time.DayOfWeek.FRIDAY;
import static java.time.DayOfWeek.MONDAY;
import static java.time.DayOfWeek.THURSDAY;
import static java.time.DayOfWeek.TUESDAY;
import static java.time.DayOfWeek.WEDNESDAY;
import static org.assertj.core.api.Assertions.assertThat;

import com.heybys.oddments.fooddelivery.domain.generic.Money;
import com.heybys.oddments.fooddelivery.domain.generic.TimePeriod;
import com.heybys.oddments.fooddelivery.domain.shop.MenuRepository;
import com.heybys.oddments.fooddelivery.domain.shop.Shop;
import com.heybys.oddments.fooddelivery.domain.shop.ShopId;
import com.heybys.oddments.fooddelivery.domain.shop.ShopRepository;
import java.time.DayOfWeek;
import java.time.LocalTime;
import java.util.HashMap;
import java.util.Map;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;

@Transactional
@SpringBootTest
class ShopServiceTest {

    private static Shop shop;

    private static ShopId shopId;

    private final ShopRepository shopRepository;

    private final MenuRepository menuRepository;

    private final ShopService shopService;

    @Autowired
    public ShopServiceTest(MenuRepository menuRepository, ShopRepository shopRepository) {
        this.menuRepository = menuRepository;
        this.shopRepository = shopRepository;
        this.shopService = new ShopService(menuRepository, shopRepository);
    }

    @BeforeAll
    static void beforeAll() {
        Map<DayOfWeek, TimePeriod> operatingHours = new HashMap<>();
        operatingHours.put(MONDAY, TimePeriod.between(LocalTime.of(9, 0), LocalTime.of(18, 0)));
        operatingHours.put(TUESDAY, TimePeriod.between(LocalTime.of(9, 0), LocalTime.of(18, 0)));
        operatingHours.put(WEDNESDAY, TimePeriod.between(LocalTime.of(9, 0), LocalTime.of(18, 0)));
        operatingHours.put(THURSDAY, TimePeriod.between(LocalTime.of(9, 0), LocalTime.of(18, 0)));
        operatingHours.put(FRIDAY, TimePeriod.between(LocalTime.of(9, 0), LocalTime.of(18, 0)));

        shopId = new ShopId(1L);

        shop = new Shop(shopId, "Test", Money.wons(32L), operatingHours);
    }

    @DisplayName("put_off_one_hour_on_successfully")
    @Test
    void putOffOneHourOn() {
        // given
        shopRepository.add(shop);
        TimePeriod oldFridayTimePeriod = shop.getOperatingHours().get(FRIDAY);

        // when
        shopService.putOffOneHourOn(shopId, FRIDAY);

        // then
        Shop found = shopRepository.find(shopId);
        TimePeriod fridayTimePeriod = found.getOperatingHours().get(FRIDAY);

        assertThat(fridayTimePeriod).isEqualTo(oldFridayTimePeriod.putOffHours(1));
    }
}

File: src/test/kotlin/com/heybys/oddments/fooddelivery/service/shop/ShopServiceTest.kt
package com.heybys.oddments.fooddelivery.service.shop

import com.heybys.oddments.fooddelivery.domain.generic.Money
import com.heybys.oddments.fooddelivery.domain.generic.TimePeriod
import com.heybys.oddments.fooddelivery.domain.shop.MenuRepository
import com.heybys.oddments.fooddelivery.domain.shop.Shop
import com.heybys.oddments.fooddelivery.domain.shop.ShopId
import com.heybys.oddments.fooddelivery.domain.shop.ShopRepository
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.transaction.annotation.Transactional
import java.time.DayOfWeek.FRIDAY
import java.time.DayOfWeek.MONDAY
import java.time.DayOfWeek.THURSDAY
import java.time.DayOfWeek.TUESDAY
import java.time.DayOfWeek.WEDNESDAY
import java.time.LocalTime

@Transactional
@SpringBootTest
class ShopServiceTest
@Autowired
constructor(
    private val menuRepository: MenuRepository,
    private val shopRepository: ShopRepository,
) {

    private val shopService = ShopService(menuRepository, shopRepository)

    companion object {
        private lateinit var shop: Shop
        private lateinit var shopId: ShopId

        @JvmStatic
        @BeforeAll
        fun beforeAll() {
            val operatingHours =
                mutableMapOf(
                    MONDAY to TimePeriod.between(LocalTime.of(9, 0), LocalTime.of(18, 0)),
                    TUESDAY to TimePeriod.between(LocalTime.of(9, 0), LocalTime.of(18, 0)),
                    WEDNESDAY to TimePeriod.between(LocalTime.of(9, 0), LocalTime.of(18, 0)),
                    THURSDAY to TimePeriod.between(LocalTime.of(9, 0), LocalTime.of(18, 0)),
                    FRIDAY to TimePeriod.between(LocalTime.of(9, 0), LocalTime.of(18, 0)),
                )

            shopId = ShopId(1L)
            shop = Shop(shopId, "Test", Money.wons(32L), operatingHours)
        }
    }

    @DisplayName("put_off_one_hour_on_successfully")
    @Test
    fun putOffOneHourOn() {
        // given
        shopRepository.save(shop)
        val oldFridayTimePeriod = shop.operatingHours?.get(FRIDAY)

        // when
        shopService.putOffOneHourOn(shopId, FRIDAY)

        // then
        val found = shopRepository.find(shopId)
        val fridayTimePeriod = found.operatingHours?.get(FRIDAY)

        assertThat(fridayTimePeriod).isEqualTo(oldFridayTimePeriod?.putOffHours(1))
    }
}
